### DISCUSSION TITLE
执行sql时，报列不存在或者表不存在怎么办

### DISCUSSION THREAD

**Discussion by PanternBao** [actiontech/dble#2837] [Q&A] [answered]
在执行sql的过程中，有时候会出现下面的报错：

> table meta 'xxxx' is lost, pelase reload @@metadata

或者

执行的查询语句中的字段在mysql库表里面真实存在，但是在dble中查询这个列有时候可以执行，有时候报列不存在
比如：
> select .... c1 from t1 ... order by c2 这样报找不到字段C1

但是

> select .... c1 from t1 ... 不排序的话不会报找不到字段

遇到这样的问题怎么排查？

**Comment by PanternBao** [comment, marked as answer]
上述的问题其实都是和dble的表元数据相关，首先得搞明白dble中表元数据的相关概念。
这里我将dble中语句简单的分为两种类型：
- 可直接下推mysql执行
- 需要dble解析处理

#### 直接下推

对于此类语句，dble仅仅将sql下推到mysql节点上执行，并且转发mysql的结果

#### 需要dble解析处理

对于此类语句，dble会对sql进行解析，检查sql中涉及表的列是否存在，此时会用到dble中的表元数据信息
对于mysql返回的结果，dble不仅仅是简单的转发，会进行排序，去重等复杂的操作。

#### 如何区分两种类型
通过dble提供的explain来区分

下面的例子是直接下推的sql示例：

> mysql> explain select * from global1 t1 left join sharding_2_t1 t2 on t1.id=t2.id where t1.id=2;
> +---------------+----------+----------------------------------------------------------------------------------+
> | SHARDING_NODE | TYPE     | SQL/REF                                                                          |
> +---------------+----------+----------------------------------------------------------------------------------+
> | dn1           | BASE SQL | select * from global1 t1 left join sharding_2_t1 t2 on t1.id=t2.id where t1.id=2 |
> +---------------+----------+----------------------------------------------------------------------------------+
> 1 row in set (0.04 sec)

这样就很容易解释：
`select .... c1 from t1 ... order by c2` 这样报找不到字段c1（这个语句不会直接下推，需要dble解析处理）
但是
`select .... c1 from t1 ...`不排序的话就没问题 不会报找不到字段（这个语句会直接下推）

#### 问题：
dble中的表元数据是从mysql中获取的，因此会存在不一致的情形，问题中的两种情况就比较典型
出现此类错误，需要自检一下做ddl的规范性：
- 是否通过dble执行ddl语句
- 仅仅在mysql节点上执行ddl语句，而没有通过管理端命令同步到dble的元数据中

### 如何解决元数据不同步
1. 可以先通过check metadata命令检查dble中的元数据
https://github.com/actiontech/dble-docs-cn/blob/master/2.Function/2.1_manager_cmd/2.1.17_check_meta_data.md#21170--check-metadata-%E5%91%BD%E4%BB%A4

2. 同步元数据

> reload @@metadata

理论上，reload @@config 也会有同步元数据的功能，但是，reload @@config更加智能，在配置没有变更的情况下，不会同步元数据，因此可以直接使用reload @@metadata

**Reply by yanhuqing666** [reply]
reload @@metadata 不是集群命令，需要每个节点都执行一遍
